js/private/js_library.bzl

```diff
@@ -68,48 +83,84 @@ target in {file_basename}'s package and add that target to the deps of {this_tar
+            declarations.append(src)
+        elif src.path.endswith("/package.json"):
+            # package.json may be required to resolve "typings" key
+            declarations.append(src)
+            sources.append(src)
+        else:
+            sources.append(src)
+
+    sources_depset = depset(sources)
+    transitive_sources_depset = depset(
+        direct = sources,
+        transitive = [
+            target[JsInfo].transitive_sources
+            for target in ctx.attr.deps
+            if JsInfo in target
+        ],
+    )
 
-    runfiles = ctx.runfiles(
-        files = output_srcs,
-        # We do not include typings_depsets in the runfiles because that would cause type-check actions to occur
-        # in every development workflow.
-        transitive_files = depset(transitive = files_depsets),
+    declarations_depset = depset(declarations)
+    transitive_declarations_depset = depset(
+        direct = declarations,
+        transitive = [
+            target[DeclarationInfo].transitive_declarations
+            for target in ctx.attr.deps
+            if DeclarationInfo in target
+        ],
     )
-    deps_runfiles = [d[DefaultInfo].default_runfiles for d in ctx.attr.deps]
+
+    npm_linked_deps_depsets = [
+        target[JsInfo].npm_linked_deps
+        for target in ctx.attr.deps
+        if JsInfo in target
+    ] + [
+        target[NpmLinkedPackageInfo].transitive_files
+        for target in ctx.attr.deps
+        if NpmLinkedPackageInfo in target
+    ]
+    for target in ctx.attr.deps:
+        if NpmLinkedPackagesInfo in target:
+            for package in target[NpmLinkedPackagesInfo].packages:
+                npm_linked_deps_depsets.append(package.transitive_files)
+    npm_linked_deps_depset = depset(transitive = npm_linked_deps_depsets)
 
     return [
+        JsInfo(
+            sources = sources_depset,
+            transitive_sources = transitive_sources_depset,
+            declarations = declarations_depset,
+            transitive_declarations = transitive_declarations_depset,
+            npm_linked_deps = npm_linked_deps_depset,
+            # npm_link_package_deps = ,
+        ),
         DefaultInfo(
-            files = depset(transitive = files_depsets),
-            runfiles = runfiles.merge_all(deps_runfiles),
+            files = transitive_sources_depset,
         ),
-        declaration_info(
-            declarations = depset(typings),
-            deps = ctx.attr.deps,
+        DeclarationInfo(
+            declarations = declarations_depset,
+            transitive_declarations = transitive_declarations_depset,
         ),
-        OutputGroupInfo(types = typings),
+        OutputGroupInfo(types = declarations_depset),
     ]
 
 js_library_lib = struct(
     attrs = _ATTRS,
     implementation = _js_library_impl,
-    provides = [DefaultInfo, DeclarationInfo],
+    provides = [DefaultInfo, DeclarationInfo, JsInfo, OutputGroupInfo],
```

Comment:
`OutputGroupInfo` was missing? Does that mean we have no tests for it and should create a task to add tests?

Reply:
I think that `OutputGroupInfo` and `DefaultInfo` are always implied and don't need to be specified